### DISCUSSION TITLE
Fetch Audit Data for Navigation Properties

### DISCUSSION THREAD

**Discussion by manideep418** [thepirat000/Audit.NET#383] [Q&A]
Hello,

I have the junction table with many-to-many relation like below

`Student -> StudentCourse <- Course
`
And my AuditCore config is like below 

Audit.Core.Configuration.Setup()
	.UseEntityFramework(_ => _
		.AuditTypeMapper(t => typeof(AuditLog))
        .AuditEntityAction<AuditLog>((ev, entry, entity) =>
        {
			entity.AuditData = entry.ToJson();
            entity.EntityType = entry.EntityType.Name;
            entity.AuditDate = DateTimeOffset.Now;
            entity.AuditUser = ev.CustomFields["ContextUser"].ToString();
        })
        .IgnoreMatchedProperties(true));

Now I would like to get the AuditData from Audit Event even from the NavigationProperty i.e., StudentCourse. Could you please help me with any suggestions.

**Comment by thepirat000** [comment]
Not sure if I understand, but you can obtain the EF audit event with `GetEntityFrameworkEvent` extension on `AuditEvent`

```c#
using Audit.EntityFramework;
...

var efEvent = auditEvent.GetEntityFrameworkEvent();

string table = efEvent.Entries[0].Table;
obejct original_entity = efEvent.Entries[0].GetEntry().Entity;
//...
```

Or if you want to access the audited entity object on the action:

```c#
.AuditEntityAction((ev, entry, auditEntity) =>
{
  var entity = entry.GetEntry().Entity;
}
```

**Reply by manideep418** [reply]
In the above example entry.ToJson() only gives the EF core Entry only for the Student Table. But I would like to get the EF core Entry(current and old values) for StudentCourse table.

Basically I would like to achieve this in EF core 5.. but its not working..
```
IncludeIndependantAssociations: Value to indicate if the Independant Associations should be included. Default is false. (Only for EF <= 6.2)
Note: EF Core <= 3 does not support many-to-many relations without a join entity, and for EF Core 5 the many-to-many relations are normally included on the audit event entries.
```

**Reply by thepirat000** [reply]
If you could provide an example of what are you changing on the dbcontext and what are your expectations that would be fine.

Note the last version (16.5.5) released yesterday, includes a new mapping configuration for the EF Data Provider (MapTable / MapExplicit) that allows to audit tables that are not mapped to a type, like the many-to-many implicitly created tables.

Check the example https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#map-many-to-many-relations-without-join-entity

If I'm understanding correctly, your case is slightly different since all your audits go to the same `AuditLog` entity. So you could change your configuration to something like this:

```c#
Audit.Core.Configuration.Setup()
    .UseEntityFramework(_ => _
    .AuditTypeExplicitMapper(exp => exp
        .MapExplicit<AuditLog>(entry => true)   // Maps all the entries to AuditLog, including entries not mapped to a type
        .AuditEntityAction((ev, entry, entity) =>
        {
            entity.AuditData = entry.ToJson();
            entity.EntityType = entry.EntityType.Name;
            entity.AuditDate = DateTimeOffset.Now;
            entity.AuditUser = ev.CustomFields["ContextUser"].ToString();
        }))
    .IgnoreMatchedProperties(true));

```

**Reply by manideep418** [reply]
As I said above.. I would like to store the Junction table audit history on the main entity creation or update.
The DB context change I am doing is adding the Audience Entity with PostalCodes in the

`_dbContext.Audience.Add(aud);`


For Example.. I have an Audience Entity, with properties
```
Name,
ID,
IEnumerable< PostalCodes > => AudiencePostalCode.Select(pc => pc.PostalCode)
```

```
public partial class Audience 
    {
        public Audience()
        {
            AudiencePostalCode = new HashSet<AudiencePostalCode>();
        }

        public int AudienceId { get; set; }
        public string Name { get; set; }
                public string CreateUser { get; set; }
        public string UpdateUser { get; set; }

        public IEnumerable<CPostalCodes= >> AudienceCountry.Select(c => c.Country);
```

So whenever I create an Audience.. I do provide Postal Codes.. SO I want to store the AUdiencePostalCode entity within Audience.. Instead of having it as separate row in AuditLog Table.

**Reply by thepirat000** [reply]
If it's part of the original entity, you can just enable the [`IncludeEntityObjects`](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#settings) setting, and the entity object will be included on `entry.Entity`

**Reply by manideep418** [reply]
How can I do IncludeEntityObjects only for Audience but not for all other Entities in the DBCOntext?

**Reply by thepirat000** [reply]
That's not currently possible. since it appies to the entire dbcontext.
But you have options:
- Include entity objects but remove the property before saving
```c#
.AuditEntityAction((ev, entry, entity) =>
{
    if (entry.Table != "Audience") 
        entry.Entity = null; 
    // ...
}
```
- Exclude entity objects, but get the entity and include it before saving
```c#
.AuditEntityAction((ev, entry, entity) =>
{
    if (entry.Table == "Audience") 
        entry.Entity = entry.GetEntry().Entity; 
    // ...
}
```
